libyul/optimiser/MemoryEscalator.cpp

```diff
@@ -0,0 +1,137 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/MemoryEscalator.h>
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/FunctionCallFinder.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/AsmData.h>
+#include <libyul/CompilabilityChecker.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+#include <libyul/Object.h>
+#include <libyul/Utilities.h>
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+#include <list>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+
+void MemoryEscalator::run(OptimiserStepContext& _context, Object& _object, bool _optimizeStackAllocation)
+{
+	if (!_object.code)
+		return;
+
+	// Determine which variables need to be moved.
+	auto functionStackErrorInfo = CompilabilityChecker::run(_context.dialect, _object, _optimizeStackAllocation);
+	if (functionStackErrorInfo.empty())
+		return;
+
+	// Find the literal argument of the ``memoryinit`` call, if there is a unique such call, otherwise abort.
+	yulAssert(dynamic_cast<EVMDialect const*>(&_context.dialect), "MemoryEscalator can only be run on EVMDialect objects.");
+	Literal* memoryInitLiteral = nullptr;
+	if (
+		auto memoryInits = FunctionCallFinder::run(*_object.code, "memoryinit"_yulstring);
+		memoryInits.size() == 1
+	)
+		memoryInitLiteral = std::get_if<Literal>(&memoryInits.front()->arguments.back());
+	if (!memoryInitLiteral)
+		return;
+	u256 reservedMemory = valueOfLiteral(*memoryInitLiteral);
+
+	auto callGraph = CallGraphGenerator::callGraph(*_object.code).functionCalls;
+
+	// Collect all names of functions contained in cycles in the callgraph.
```

Comment:
`util::CycleDetector`? Semantics.cpp contains the same code by the way.

Reply:
Maybe... I thought about using ``CycleDetector``, but came to the conclusion that it's not really well suited for this (containing a cycle is not exactly being contained in a cycle) - and I'd need to have a second look, but it may be that what Semantics.cpp does is even worse complexity-wise than this here :-).

I'd call this a placeholder for now and would think about improving it once everything else is fine. Maybe in the end I'll just look up an optimal algorithm for this and plug it in here and into Semantics.cpp...